### PR TITLE
[metricbeat] migrate mongodb to reporterv2 with error handling

### PR DESCRIPTION
**Pull request by fearful-symmetry**
See elastic/beats#11374 

Not much to add to this one.

### CODE REVIEW

File: metricbeat/module/mongodb/collstats/collstats.go [hunk 1]
```diff
@@ -99,22 +85,18 @@ func (m *MetricSet) Fetch(reporter mb.ReporterV2) {
 
 		infoMap, ok := info.(common.MapStr)
 		if !ok {
-			err = errors.New("Unexpected data returned by mongodb")
-			logger.Error(err)
-			reporter.Error(err)
+			m.Logger().Error("Unexpected data returned by mongodb")
```
Comment (ruflin) [inline metricbeat/module/mongodb/collstats/collstats.go]: This is a change in behaviour. Previously we also sent an error event now we only log it. I think we should still send the error event.

File: metricbeat/module/mongodb/collstats/collstats.go [hunk 2]
```diff
@@ -99,22 +85,18 @@ func (m *MetricSet) Fetch(reporter mb.ReporterV2) {
...
 			continue
 		}
 
 		event, err := eventMapping(group, infoMap)
 		if err != nil {
-			err = errors.Wrap(err, "Mapping of the event data filed")
-			logger.Error(err)
-			reporter.Error(err)
+			m.Logger().Error(errors.Wrap(err, "Mapping of the event data filed"))
```
Comment (ruflin) [inline metricbeat/module/mongodb/collstats/collstats.go]: Same as above.

File: metricbeat/module/mongodb/dbstats/dbstats.go [hunk 3]
```diff
@@ -87,23 +82,23 @@ func (m *MetricSet) Fetch(reporter mb.ReporterV2) {
 
 		err := db.Run("dbStats", &result)
 		if err != nil {
-			err = errors.Wrapf(err, "Failed to retrieve stats for db %s", dbName)
-			logger.Error(err)
-			continue
+			m.Logger().Error(errors.Wrapf(err, "Failed to retrieve stats for db %s", dbName))
```
Comment (ruflin) [inline metricbeat/module/mongodb/dbstats/dbstats.go]: We lost the `continue` here and are not reporting the error event anymore.

File: metricbeat/module/mongodb/dbstats/dbstats.go [hunk 4]
```diff
@@ -87,23 +82,23 @@ func (m *MetricSet) Fetch(reporter mb.ReporterV2) {
...
+
 		}
 		data, _ := schema.Apply(result)
-		if reported := reporter.Event(mb.Event{MetricSetFields: data}); !reported {
-			logger.Debug("error reporting event")
-			return
+		reported := reporter.Event(mb.Event{MetricSetFields: data})
+		if !reported {
+			//this means the metricset has closed, no point in trying to log anything further
+			return errors.New("metricset has stopped")
```
Comment (ruflin) [inline metricbeat/module/mongodb/dbstats/dbstats.go]: I think here we should just `return nil` and not report an error document. Stopping a metricset is not an error.

File: metricbeat/module/mongodb/dbstats/dbstats.go (hunk 3 above)
Comment (fearful-symmetry) [reply metricbeat/module/mongodb/dbstats/dbstats.go]: Good catch!

File: metricbeat/module/mongodb/collstats/collstats.go [hunk 5]
```diff
@@ -57,39 +54,28 @@ func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
 // Fetch methods implements the data gathering and data conversion to the right
 // format. It publishes the event which is then forwarded to the output. In case
 // of an error set the Error field of mb.Event or simply call report.Error().
-func (m *MetricSet) Fetch(reporter mb.ReporterV2) {
+func (m *MetricSet) Fetch(reporter mb.ReporterV2) error {
 	// instantiate direct connections to each of the configured Mongo hosts
 	mongoSession, err := mongodb.NewDirectSession(m.DialInfo)
 	if err != nil {
-		logger.Error(err)
-		reporter.Error(err)
```
Comment (kaiyan-sheng) [inline metricbeat/module/mongodb/collstats/collstats.go]: This is still missing here 😄 same in line 76.

Comment (fearful-symmetry) [reply metricbeat/module/mongodb/collstats/collstats.go]: Ah, missed one!

Comment (fearful-symmetry) [reply metricbeat/module/mongodb/collstats/collstats.go]: Wait, do we need to add anything there? My understanding is that we only need to add the error reporting back in where we're not actually returning, i.e, inside a loop

Comment (kaiyan-sheng) [reply metricbeat/module/mongodb/collstats/collstats.go]: My understanding is `logger.Error(err)` is to log the error and `reporter.Error(err)` is to actually report an event with error msg. So if we only have a `return errors.Wrap(...)` here, then we are missing the report an event part.

Comment (ruflin) [reply metricbeat/module/mongodb/collstats/collstats.go]: If we return an error, it's not only logged but also reported: https://github.com/elastic/beats/blob/master/metricbeat/mb/module/wrapper.go#L243

File: metricbeat/module/mongodb/dbstats/dbstats.go [hunk 6]
```diff
@@ -88,22 +83,22 @@ func (m *MetricSet) Fetch(reporter mb.ReporterV2) {
...
+			reporter.Error(err)
+			m.Logger().Error(err)
 			continue
 		}
 		data, _ := schema.Apply(result)
-		if reported := reporter.Event(mb.Event{MetricSetFields: data}); !reported {
-			logger.Debug("error reporting event")
-			return
+		reported := reporter.Event(mb.Event{MetricSetFields: data})
+		if !reported {
```
Comment (ruflin) [inline metricbeat/module/mongodb/dbstats/dbstats.go]: We could log on debug level that the metricset was stopped during reporting, but I think that can already be seen in other places, so +1 on this change.

Nit: I preferred the previous 1 line statement ;-)

**Pull request merged**